### ISSUE TITLE
Ternary term generates cuda errors and warnings

### ISSUE THREAD

**Issue by trypag** [facebookresearch/TensorComprehensions#144]
Hi,
- OS: Fedora 25
- How you installed TC (docker, conda, source): conda
- Python version: 3.6
- CUDA/cuDNN version: 8.0
- Conda version (if using conda): 4.4.11

I am trying to write the backward of this TC, the forward works as expected, but autotuning the backward generates weird cuda errors that are impossible to understand.
I noticed that removing the ternary term of the second line in backward will solve the issue, I guess this is the root cause.
Thanks for your help.

```
import tensor_comprehensions as tc
import torch


norm_lang = """
def norm_max(float(N, C, H, W) I) -> (O, P, maxVal) {{
    O(n, c, h, w) = exp(I(n, c, h, w) * {beta})
    maxVal(n, h, w) max= O(n, c, h, w)
    P(n, c, h, w) = O(n, c, h, w) / maxVal(n, h, w)
}}

def norm_max_grad(float(N, C, H, W) I, float(N, C, H, W) O_grad, float(N, C, H, W) P_grad, float(N, H, W) maxVal) -> (I_grad){{
    I_grad(n, c, h, w) +=! exp(I(n, c, h, w) * {beta}) * {beta} * maxVal(n, h, w)
    I_grad(n, c, h, w) = I_grad(n, c, h, w) - (exp(I(n, c, h, w) * {beta}) * (I(n, c, h, w) == maxVal(n, h, w) ? 1: 0))
    I_grad(n, c, h, w) = I_grad(n, c, h, w) / pow(maxVal(n, h, w), 2)
    I_grad(n, c, h, w) = I_grad(n, c, h, w) * I_grad(n, c, h, w)
}}
"""

norm_max = tc.define(norm_lang, training=True, name="norm_max",
                     backward="norm_max_grad",
                     constants={'beta': 1})
norm_max.autotune(torch.randn(8, 135, 184, 184).cuda(), generations=1, pop_size=10)

```

```
[WARNING]: Autotuning results won't be cached. 'cache' option is not set
[WARNING]: Using 'naive' type mapping options for autotuning. See help(your_layer.autotune) for how to set mapping options.
Generation 0    Jobs(Compiled, GPU)/total  (10, 10)/10   (best/median/worst)us: 21383/66695/86822
[INFO]: Autotuning the backward layer now
[WARNING]: Using 'naive' type mapping options for autotuning. See help(your_layer.autotune) for how to set mapping options.
WARNING: Logging before InitGoogleLogging() is written to STDERR
W0313 14:40:08.700757 14911 rtc.cc:103] Compilation failure for nvrtc(NVRTC_ERROR_COMPILATION): 
default_program(32): error: expected a ")"

default_program(32): error: expected a ")"

default_program(32): error: expected a ")"

default_program(32): error: expected an expression

default_program(18): warning: variable "b0" was declared but never referenced

default_program(18): warning: variable "b2" was declared but never referenced

default_program(19): warning: variable "t2" was declared but never referenced

default_program(22): warning: variable "O_grad" was declared but never referenced

default_program(23): warning: variable "P_grad" was declared but never referenced

4 errors detected in the compilation of "default_program".
 source:
template<typename T> inline __device__ T floord(T n, T d) {
  return n < 0 ? - (-n + d - 1)/d : n / d;
}
#define if_then_else(cond,a,b) (cond) ? (a) : (b);

// Halide type handling
typedef int int32;
typedef long int64;
typedef float float32;
typedef double float64;

#define inff __int_as_float(0x7f800000)
#define inf __longlong_as_double(0x7ff0000000000000LL)

extern "C" {
__global__ void norm_max_grad_135_184_8_184(int32 C, int32 H, int32 N, int32 W, float32* pI_grad, float32* pI, float32* pO_grad, float32* pP_grad, float32* pmaxVal) {
  int b0 = blockIdx.x; int b1 = blockIdx.y; int b2 = blockIdx.z;
  int t0 = threadIdx.x; int t1 = threadIdx.y; int t2 = threadIdx.z;
  float32 (*I_grad)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pI_grad);
  float32 (*I)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pI);
  float32 (*O_grad)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pO_grad);
  float32 (*P_grad)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pP_grad);
  float32 (*maxVal)[184][184] = reinterpret_cast<float32 (*)[184][184]>(pmaxVal);
  for (int c2 = 0; c2 <= 183; c2 += 32) {
    for (int c4 = 0; c4 <= 7; c4 += 1) {
      for (int c5 = 0; c5 <= min(31, -32 * b1 + 134); c5 += 1) {
        for (int c6 = t1; c6 <= min(31, -c2 + 183); c6 += 8) {
          for (int c7 = t0; c7 <= 183; c7 += 32) {
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = 0.000000f;
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7] + (exp(I[c4][32*b1 + c5][c2 + c6][c7])*maxVal[c4][c2 + c6][c7]));
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7] - (exp(I[c4][32*b1 + c5][c2 + c6][c7])*float32(if_then_else((I[c4][32*b1 + c5][c2 + c6][c7] == maxVal[c4][c2 + c6][c7]), 1, 0))));
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7]/pow(maxVal[c4][c2 + c6][c7], 2));
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7]*I_grad[c4][32*b1 + c5][c2 + c6][c7]);
          }
        }
      }
    }
  }
}
}

/*
Mapping Options:
tc::MappingOptions::makeNaiveMappingOptions()
    .outerScheduleFusionStrategy(tc::FusionStrategy::Preserve3Coincident)
    .outerScheduleAllowSkewing(false)
    .outerSchedulePositiveOrthant(true)
    .intraTileScheduleFusionStrategy(tc::FusionStrategy::Preserve3Coincident)
    .intraTileScheduleAllowSkewing(false)
    .intraTileSchedulePositiveOrthant(true)
    .tile(32, 32, 32)
    .mapToThreads(32, 8)
    .mapToBlocks(256, 256)
    .unroll(1)
    .tileImperfectlyNested(false)
    .useSharedMemory(false)
    .usePrivateMemory(false)
    .unrollCopyShared(false)
    .matchLibraryCalls(false);
TC version: 8e112e9dccda62c30ef29208a827e783b9a7f156
*/
E0313 14:40:08.700851 14911 rtc.cc:106] Compilation failure for nvrtc(NVRTC_ERROR_COMPILATION): 
default_program(32): error: expected a ")"

default_program(32): error: expected a ")"

default_program(32): error: expected a ")"

default_program(32): error: expected an expression

default_program(18): warning: variable "b0" was declared but never referenced

default_program(18): warning: variable "b2" was declared but never referenced

default_program(19): warning: variable "t2" was declared but never referenced

default_program(22): warning: variable "O_grad" was declared but never referenced

default_program(23): warning: variable "P_grad" was declared but never referenced

4 errors detected in the compilation of "default_program".
 source:
template<typename T> inline __device__ T floord(T n, T d) {
  return n < 0 ? - (-n + d - 1)/d : n / d;
}
#define if_then_else(cond,a,b) (cond) ? (a) : (b);

// Halide type handling
typedef int int32;
typedef long int64;
typedef float float32;
typedef double float64;

#define inff __int_as_float(0x7f800000)
#define inf __longlong_as_double(0x7ff0000000000000LL)

extern "C" {
__global__ void norm_max_grad_135_184_8_184(int32 C, int32 H, int32 N, int32 W, float32* pI_grad, float32* pI, float32* pO_grad, float32* pP_grad, float32* pmaxVal) {
  int b0 = blockIdx.x; int b1 = blockIdx.y; int b2 = blockIdx.z;
  int t0 = threadIdx.x; int t1 = threadIdx.y; int t2 = threadIdx.z;
  float32 (*I_grad)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pI_grad);
  float32 (*I)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pI);
  float32 (*O_grad)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pO_grad);
  float32 (*P_grad)[135][184][184] = reinterpret_cast<float32 (*)[135][184][184]>(pP_grad);
  float32 (*maxVal)[184][184] = reinterpret_cast<float32 (*)[184][184]>(pmaxVal);
  for (int c2 = 0; c2 <= 183; c2 += 32) {
    for (int c4 = 0; c4 <= 7; c4 += 1) {
      for (int c5 = 0; c5 <= min(31, -32 * b1 + 134); c5 += 1) {
        for (int c6 = t1; c6 <= min(31, -c2 + 183); c6 += 8) {
          for (int c7 = t0; c7 <= 183; c7 += 32) {
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = 0.000000f;
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7] + (exp(I[c4][32*b1 + c5][c2 + c6][c7])*maxVal[c4][c2 + c6][c7]));
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7] - (exp(I[c4][32*b1 + c5][c2 + c6][c7])*float32(if_then_else((I[c4][32*b1 + c5][c2 + c6][c7] == maxVal[c4][c2 + c6][c7]), 1, 0))));
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7]/pow(maxVal[c4][c2 + c6][c7], 2));
            I_grad[c4][32*b1 + c5][c2 + c6][c7] = (I_grad[c4][32*b1 + c5][c2 + c6][c7]*I_grad[c4][32*b1 + c5][c2 + c6][c7]);
          }
        }
      }
    }
  }
}
}

/*
Mapping Options:
tc::MappingOptions::makeNaiveMappingOptions()
    .outerScheduleFusionStrategy(tc::FusionStrategy::Preserve3Coincident)
    .outerScheduleAllowSkewing(false)
    .outerSchedulePositiveOrthant(true)
    .intraTileScheduleFusionStrategy(tc::FusionStrategy::Preserve3Coincident)
    .intraTileScheduleAllowSkewing(false)
    .intraTileSchedulePositiveOrthant(true)
    .tile(32, 32, 32)
    .mapToThreads(32, 8)
    .mapToBlocks(256, 256)
    .unroll(1)
    .tileImperfectlyNested(false)
    .useSharedMemory(false)
    .usePrivateMemory(false)
    .unrollCopyShared(false)
    .matchLibraryCalls(false);
TC version: 8e112e9dccda62c30ef29208a827e783b9a7f156
*/
[ERROR]: Raised exception: Could not compile function

```

**Comment by prigoyal** [comment]
thanks @trypag, let me investigate and report back :)

**Comment by ftynse** [comment]
@prigoyal I see the problem, you can stop looking

Here, somebody decided to put a semicolon in a macro, which is usually a bad idea.
`#define if_then_else(cond,a,b) (cond) ? (a) : (b);`

This macro is used in a more complex expression and the semicolon gets inserted in the middle (between the ?: expansion and the closing parens)

**Comment by prigoyal** [comment]
thanks @ftynse, I noticed that however I am trying to provide a workaround to the user since binaries won't be refreshed soon. Thanks for your help.

**Comment by ftynse** [comment]
That code is emitted, I don't see any easy way to work that around....

**Comment by prigoyal** [comment]
@trypag I have a simple workaround for you. This below runs fine:

```
import tensor_comprehensions as tc
import torch
import unittest

class TestMaxNorm(unittest.TestCase):

    def test_max_norm(self):
        norm_lang = """
        def norm_max(float(N, C, H, W) I) -> (O, P, maxVal) {{
            O(n, c, h, w) = exp(I(n, c, h, w) * {beta})
            maxVal(n, h, w) max= O(n, c, h, w)
            P(n, c, h, w) = O(n, c, h, w) / maxVal(n, h, w)
        }}
        def norm_max_grad(float(N, C, H, W) I, float(N, C, H, W) O_grad, float(N, C, H, W) P_grad, float(N, H, W) maxVal) -> (I_grad, temp) {{
            I_grad(n, c, h, w) +=! exp(I(n, c, h, w) * {beta}) * {beta} * maxVal(n, h, w)
            temp(n, c, h, w) = I(n, c, h, w) == maxVal(n, h, w) ? 1: 0
            I_grad(n, c, h, w) = I_grad(n, c, h, w) - (exp(I(n, c, h, w) * {beta}) * temp(n, c, h, w))
            I_grad(n, c, h, w) = I_grad(n, c, h, w) / pow(maxVal(n, h, w), 2)
            I_grad(n, c, h, w) = I_grad(n, c, h, w) * I_grad(n, c, h, w)
        }}
        """
        norm_max = tc.define(norm_lang, training=True, name="norm_max",
                             backward="norm_max_grad", constants={'beta': 1})
        norm_max.autotune(torch.randn(8, 135, 184, 184).cuda(), generations=1, pop_size=10)
```

**Comment by trypag** [comment]
Is it okay to have 2 returned variables in the backward as we are only expected one gradient ? that's the question I was asking lately ^^

**Comment by prigoyal** [comment]
aha I see, yeah, in the backwards, we can return the temporary variable `temp`. That is completely fine mathematically. If you have other layers on top of this one, then we can easily ignore the temp variable :)

**Comment by prigoyal** [comment]
also removing the semicolon from macro fixes this. This will be available in conda package in the next release. Until then, the `temp` variable should unblock. Let us know if something else is needed. I'll close out this task now :)
